File: Python/Pattern wise question/BFS/8_populating-next-right-pointers-in-each-node.py
"""
# Definition for a Node.
class Node:
    def __init__(self, val: int = 0, left: 'Node' = None, right: 'Node' = None, next: 'Node' = None):
        self.val = val
        self.left = left
        self.right = right
        self.next = next
"""
"""
This problem follows the Binary Tree Level Order Traversal pattern.
We can follow the same BFS approach.
The only difference is that while traversing a level we will remember the previous node to connect it with the current node.


so we will connect at each level
for that will take prevous = None
and instead of inserting we will check
if(prevous != None):
    pre.next = current # current is equal queue.popleft()
pre = current # at first time to connect

at the end
we will return root
"""
class Solution:
    def connect(self, root: 'Optional[Node]') -> 'Optional[Node]':
        if(not root):
            return None
        q = deque()
        q.append(root)

        while(q):
            sz = len(q)
            pre = None
            for i in range(sz):
                curr = q.popleft()
                if(pre != None):
                    pre.next = curr
                pre = curr
                if(curr.left):
                    q.append(curr.left)
                if(curr.right):
                    q.append(curr.right)

        return root

File: C++/Pattern wise question/BFS/8_populating-next-right-pointers-in-each-node.cpp

/*
// Definition for a Node.
class Node {
public:
    int val;
    Node* left;
    Node* right;
    Node* next;

    Node() : val(0), left(NULL), right(NULL), next(NULL) {}

    Node(int _val) : val(_val), left(NULL), right(NULL), next(NULL) {}

    Node(int _val, Node* _left, Node* _right, Node* _next)
        : val(_val), left(_left), right(_right), next(_next) {}
};
*/

class Solution {
public:
    Node* connect(Node* root) {

        if(!root)
            return nullptr;
        queue<Node*> q;
        q.push(root);
        while(!q.empty()){
            int sz = q.size();
            Node* pre = NULL;
            for(int i=0; i<sz; i++){
                Node* curr = q.front();
                q.pop();
                if(pre != NULL){
                    pre->next = curr;
                }
                pre = curr;
                if(curr->left) q.push(curr->left);

                if(curr->right) q.push(curr->right);

            }
        }
        return root;

    }
};
